### ISSUE TITLE
QuantizationMethod using and accuracy

### ISSUE THREAD

**Issue by Ming-Ivy** [sony/model_optimization#967]
### Issue Type

Performance

### Source

source

### MCT Version

1.10.0

### Describe the issue

```shell
Hello,

When I use mct.target_platform.QuantizationMethod.KMEANS in activation_quantization_method and   weights_quantization_method in int8, the reasult shows: Exception: Found 0 quantizer for target Weights that matches the requested quantization method KMEANS but there should be exactly one.The possible quantizers that were found are []. and for LUT_POT_QUANTIZER&LUT_SYM_QUANTIZER:    raise Exception(msg)
Exception: Look-Up-Table bit configuration has 8 bits, but must be less than 8

so my question is :1, why this error happen? is LUT method can only used for int4 and int2? 

2, when I apply resnet50, the accuracy for default is higher that only apply int8, just wondering about the reason or is it because mix precision not only adjust bit width but also quantization methods in each layer? 

3, I conclute the the process of mix precision like ：①Gets the float tensor to quantize
②Quantize for each layer.quantizes the float tensor in different bitwidths
③Evaluate the sensitivity (compute_distance_fn, distance_weighting_method, Hessian,etc)
④Candidate configuration for different bit width and method
⑤ILP for the best quantization config
is these right?

sorry for so many question, sincerely waiting for your reply.
```

### Code to reproduce the issue

```shell
op_qc_8bit = mct.target_platform.OpQuantizationConfig(
        activation_quantization_method=mct.target_platform.QuantizationMethod.SYMMETRIC ,
        weights_quantization_method=mct.target_platform.QuantizationMethod.UNIFORM ,
        activation_n_bits=8,
        weights_n_bits=8,
        weights_per_channel_threshold=True,
        enable_weights_quantization=True,
        enable_activation_quantization=True,
        quantization_preserving = False,#?write by myself
        fixed_scale = None,
        fixed_zero_point = None,
        weights_multiplier_nbits = 8, 
        simd_size=32
    )

    my_default_qco= mct.target_platform.QuantizationConfigOptions([op_qc_8bit], base_config=op_qc_8bit)
    my_tp_model= mct.target_platform.TargetPlatformModel(default_qco = my_default_qco, name='default_tp_model')
    my_target_platform_capabilities=mct.target_platform.TargetPlatformCapabilities(tp_model=my_tp_model, name='base', version=None)
    print('my_target_platform_capabilities!!!!!',my_target_platform_capabilities)

    # Run model compression toolkit
    quantized_model, quantization_info = quant.quantize(float_model,
                                                        ml.get_representative_dataset,
                                                        my_target_platform_capabilities,
                                                        config)
```

**Comment by elad-c** [comment]
Hi @Ming-Ivy ,

**regarding your questions**:

1. LUT Quantization
    **a.** QuantizationMethod.KMEANS will be removed soon in an upcoming PR, so better use either QuantizationMethod.LUT_POT_QUANTIZER or QuantizationMethod.LUT_SYM_QUANTIZER
    **b.** We raise an error for quantizing a weight with 8 bits (when the LUT number of bits is 8), because there's no advantage for LUT in that scenario. Please take a look at our predefined [LUT TPC](https://github.com/sony/model_optimization/tree/main/model_compression_toolkit/target_platform_capabilities/tpc_models/imx500_tpc/v1_lut).
2. Could you supply the code you use to generate the accuracies?
3. Yes, you are right (this is the high level mixed precision workflow)

**regarding your code**:
Please create a TPC following existing examples in the MCT [TPC library](https://github.com/sony/model_optimization/tree/main/model_compression_toolkit/target_platform_capabilities)

**Comment by Ming-Ivy** [comment]
Hi,

Thank you for your answer,

May I ask what is the "high level" mixed precision workflow?

For my code,I just added 'op_qc_8bit' based on main.py because I want to try no-mix precision with int8 method as:


def quantization_flow(config: Dict) -> Tuple[float, float, QuantInfo, DatasetInfo]:
    """
    This function implements the typical workflow when using MCT.
    It begins by evaluating the performance of the floating-point model.
    Next, the model is compressed using MCT quantization techniques.
    Finally, the evaluation process is repeated on the compressed model to assess its performance.

    Args:
        config (Dict): Configurations dictionary that contains the settings for the quantization flow

    Returns:
        float_results (float): The accuracy of the floating point model
        quant_results (float): The accuracy of the quantized model
        quantization_info (QuantInfo): Information of the model optimization process from MCT
        dataset_info (DatasetInfo): Information about the evaluated dataset
    """

    # Find and import the required modules for the models collection library ("model_library")
    model_lib_module, quant_module = find_modules(config[MODEL_LIBRARY])
    model_lib = importlib.import_module(model_lib_module)
    quant = importlib.import_module(quant_module)

    # Create a ModelLibrary object that combines the selected pre-trained model with the corresponding performance
    # evaluation functionality
    ml = model_lib.ModelLib(config)

    # Get the floating point model
    float_model = ml.get_model()

    # Evaluate the float model
    float_results, _ = ml.evaluate(float_model)

    # Select the target platform capabilities (https://github.com/sony/model_optimization/blob/main/model_compression_toolkit/target_platform_capabilities/README.md)
    
    op_qc_8bit = mct.target_platform.OpQuantizationConfig(
        activation_quantization_method=mct.target_platform.QuantizationMethod.UNIFORM ,
        weights_quantization_method=mct.target_platform.QuantizationMethod.SYMMETRIC ,
        activation_n_bits=8,
        weights_n_bits=8,
        weights_per_channel_threshold=True,
        enable_weights_quantization=True,
        enable_activation_quantization=True,
        quantization_preserving = False,#
        fixed_scale = None,
        fixed_zero_point = None,
        weights_multiplier_nbits = 8, 
        simd_size=32
    )
    op_qc_4bit = mct.target_platform.OpQuantizationConfig(
        activation_quantization_method=mct.target_platform.QuantizationMethod.POWER_OF_TWO,
        weights_quantization_method=mct.target_platform.QuantizationMethod.SYMMETRIC,
        activation_n_bits=8,
        weights_n_bits=4,
        weights_per_channel_threshold=True,
        enable_weights_quantization=True,
        enable_activation_quantization=True,
        quantization_preserving = False,#
        fixed_scale = None,
        fixed_zero_point = None,
        weights_multiplier_nbits = 8, 
        simd_size=64
    )
    op_qc_2bit = mct.target_platform.OpQuantizationConfig(
        activation_quantization_method=mct.target_platform.QuantizationMethod.POWER_OF_TWO,
        weights_quantization_method=mct.target_platform.QuantizationMethod.SYMMETRIC,
        activation_n_bits=8,
        weights_n_bits=2,
        weights_per_channel_threshold=True,
        enable_weights_quantization=True,
        enable_activation_quantization=True,
        quantization_preserving = False,#
        fixed_scale = None,
        fixed_zero_point = None,
        weights_multiplier_nbits = 8,  
        simd_size=128
    )
    #mixed_precision_configuration_options = 
    mct.target_platform.QuantizationConfigOptions([op_qc_8bit,op_qc_4bit,op_qc_2bit],base_config=op_qc_8bit)

    target_platform_cap = quant.get_tpc(config[TARGET_PLATFORM_NAME], config[TARGET_PLATFORM_VERSION])
    my_default_qco= mct.target_platform.QuantizationConfigOptions([op_qc_8bit], base_config=op_qc_8bit)
    my_tp_model= mct.target_platform.TargetPlatformModel(default_qco = my_default_qco, name='default_tp_model')
    my_target_platform_capabilities=mct.target_platform.TargetPlatformCapabilities(tp_model=my_tp_model, name='base', version=None)

    # Run model compression toolkit
    quantized_model, quantization_info = quant.quantize(float_model,
                                                        ml.get_representative_dataset,
                                                        my_target_platform_capabilities,
                                                        config)

    # Evaluate quantized model
    quant_results, dataset_info = ml.evaluate(quantized_model)

    size_mb = model_size_in_mb(float_model)
    size_q_mb = model_size_in_mb(quantized_model)
    print(f'Model size!!!: {size_mb:.2f} MB')
    print(f'Model after quantization size!!!!!!: {size_q_mb:.2f} MB')
    tflite_file_path = '/data1/work/Ming/MCT/model_optimization/tutorials/quick_start/modelqq!.tflite' # Path of exported model
    mct.exporter.keras_export_model(model=quantized_model, save_model_path=tflite_file_path,
                                    serialization_format=mct.exporter.KerasExportSerializationFormat.TFLITE, quantization_format=mct.exporter.QuantizationFormat.FAKELY_QUANT)
    keras_file_path = '/data1/work/Ming/MCT/model_optimization/tutorials/quick_start/modelqq!.h5' # Path of exported model
    mct.exporter.keras_export_model(model=quantized_model, save_model_path=keras_file_path)

    return float_results, quant_results, quantization_info, dataset_info

**Comment by elad-c** [comment]
By "_high level_" I meant you described the mixed precision process correctly.

**Regrding the TPC**:
The default TPC is mixed precision "ready", but it will not activate the mixed precision if you don't ask the MCT to do it (by setting a `mixed_precision_config` in the `CoreConfig`, and setting a `target_kpi`). Therefore, it seems there's no need to edit the TPC.

**Comment by Ming-Ivy** [comment]
> By "_high level_" I meant you described the mixed precision process correctly.
> 
> **Regrding the TPC**: The default TPC is mixed precision "ready", but it will not activate the mixed precision if you don't ask the MCT to do it (by setting a `mixed_precision_config` in the `CoreConfig`, and setting a `target_kpi`). Therefore, it seems there's no need to edit the TPC.

Hi, 
Thank you for your reply o(*￣▽￣*)ブ

Do you mean that by runing the default main.py, it does not use mix precision method? if so, would you please give me an example about how to activate the mix precision? Thank you so much.

**Comment by elad-c** [comment]
Yes.

The TPC defines the bit-width your hardware supports for quantizing the weights in your model (default: bit-width of 2, 4 or 8 bits), but if you don't define a target [KPI](https://sony.github.io/model_optimization/docs/api/experimental_api_docs/modules/mixed_precision_quantization_config.html#model_compression_toolkit.KPI) for [MCT](https://github.com/sony/model_optimization/blob/main/model_compression_toolkit/ptq/keras/quantization_facade.py#L43), it will select the maximum bit-width for maximum accuracy.

If you want to apply mixed-precision, here's a simple example [notebook](https://github.com/sony/model_optimization/blob/main/tutorials/notebooks/keras/ptq/example_keras_mobilenet_mixed_precision.ipynb). Note the in that notebook, the KPI is to achieve a 25% reduction in parameter size.

**Comment by ofirgo** [comment]
Since there haven't been any recent updates or responses, I'm closing this thread. If you encounter any further issues or have additional questions regarding this topic, please feel free to reopen the issue.
